wporg-two-factor.php

```diff
@@ -33,3 +38,114 @@ function two_factor_providers( array $providers ) : array {
+ *     1: To protect against code that calls `is_super_admin()` directly, instead of checking capabilities.
+ *     2: To avoid the code in `has_cap()` that allows Super Admins to do anything unless `do_not_allow` is set.
+ *        That would interfere with reducing their capabilities to a Subscriber in `remove_capabilities_until_2fa_enabled()`.
+ */
+function remove_super_admins_until_2fa_enabled() : void {
+	global $super_admins;
+
+	$user     = wp_get_current_user();
+	$position = array_search( $user->user_login, $super_admins, true );
+
+	if ( false === $position ) {
+		return;
+	}
+
+	if ( user_requires_2fa( $user ) && ! Two_Factor_Core::is_user_using_two_factor( $user->ID ) ) {
+		unset( $super_admins[ $position ] );
+	}
+}
+
+/**
+ * Remove capabilities when a user with elevated privileges hasn't enabled 2FA.
+ *
+ * That is necessary even though we'll redirect all requests to their profile, because otherwise they could still
+ * perform privileged actions on the front end, via the REST API, etc.
+ */
+function remove_capabilities_until_2fa_enabled( array $allcaps, array $caps, array $args, WP_User $user ) : array {
+	if ( 0 === $user->ID || ! user_requires_2fa( $user ) ) {
+		return $allcaps;
+	}
+
+	if ( ! Two_Factor_Core::is_user_using_two_factor( $user->ID ) ) {
+		// This also relies on `remove_super_admins_until_2fa_enabled()`, see notes in that function.
+		$allcaps = array(
+			'subscriber' => true,
+			'read'       => true,
+		);
+
+		// The redirect will short-circuit `enable_2fa_notice()` on the initial request, but this will be called
+		// again after the redirect, and `enable_2fa_notice()` will be executed then.
+		add_action( 'current_screen', __NAMESPACE__ . '\redirect_to_2fa_settings' );
+		add_action( 'admin_notices', __NAMESPACE__ . '\render_enable_2fa_notice' );
+	}
+
+	return $allcaps;
+}
+
+/**
+ * Check if the user has enough elevated privileges to require 2FA.
+ */
+function user_requires_2fa( WP_User $user ) : bool {
+	global $supes, $trusted_deputies, $wcorg_subroles;
+
+	$required = false;
+
+	// Only checking `$supes` because a user should never be in `$super_admins` without first being in `$supes`.
+	// A user still requires 2FA even if `remove_super_admins_until_2fa_enabled()` has removed them from `$GLOBALS['super_admins']`.
+	if ( $supes && in_array( $user->user_login, $supes, true ) ) {
+		$required = true;
+	} elseif ( $trusted_deputies && in_array( $user->ID, $trusted_deputies, true ) ) {
+		$required = true;
+	} elseif ( $wcorg_subroles && array_key_exists( $user->ID, $wcorg_subroles ) ) {
+		$required = true;
+	}
+
+	return $required;
+}
+
+/**
+ * Redirect a user to their 2FA settings if they need to enable it.
+ *
+ * This isn't usually necessary, since WordPress will prevent Subscribers from visiting other Core screens, but
+ * sometimes plugins add screens that are available to Subscribers (either intentionally or not).
```

Comment:
Given they have no caps, I'm not sure this is really something that needs to be concerned with.

If a admin page/panel is enabled for everyone, including subscribers, that's hardly something we need to enforce 2FA for a super-admin for.

To me this just feels like it's going to result in some kind of buggy behaviour at some point.. 

I would personally implement this as the cap-removal, and a post-login error page / forced-redirect to profile with the "You must enable 2FA" message.

Not saying we should remove this, just suggesting that it might not strictly be needed.

Reply:
🤔 So you're saying the difference is essentially:

```diff
- Login > redirect to profile > try to visit another page > redirect to profile
+ Login > redirect to profile > try to visit another page > let them access it
```

? That seems fine to me 👍🏻 I'm more concerned about the UX, and agree that viewing pages available to subscribers isn't a big deal.

Reply:
Changed in 55de8ab29046e54c2091e31bfee35758f21a7515